src/lib/homestore_backend/hs_shard_manager.cpp

```diff
@@ -68,187 +66,192 @@ ShardManager::Result< ShardInfo > HSHomeObject::_create_shard(pg_id_t pg_owner,
+    // but do we really need to handle this for create_shard or seal_shard?
+    // because this raft log had been commit before HO restarts and commit result is already saved in HS metablks
+    // when HS restarts, all PG/Shard infos will be recovered from HS metablks and if we commit again, it will cause
+    // duplication;
 }
 
-void HSHomeObject::on_shard_message_commit(int64_t lsn, sisl::blob const& header, sisl::blob const& key,
+void HSHomeObject::do_shard_message_commit(int64_t lsn, ReplicationMessageHeader& header,
+                                           homestore::MultiBlkId const& blkids, sisl::blob value,
                                            cintrusive< homestore::repl_req_ctx >& hs_ctx) {
-    // ctx will be nullptr when:
-    // 1. on the follower side
-    // 2. on the leader side but homeobject restarts and replay all commited log entries from the last commit id;
     repl_result_ctx< ShardManager::Result< ShardInfo > >* ctx{nullptr};
     if (hs_ctx != nullptr) {
         ctx = boost::static_pointer_cast< repl_result_ctx< ShardManager::Result< ShardInfo > > >(hs_ctx).get();
     }
 
-    std::string shard_msg;
-    if (!precheck_and_decode_shard_msg(lsn, header, key, &shard_msg)) {
-        if (ctx) { ctx->promise_.setValue(folly::makeUnexpected(ShardError::INVALID_ARG)); }
+    if (header.corrupted()) {
+        LOGW("replication message header is corrupted with crc error, lsn:{}", lsn);
+        if (ctx) { ctx->promise_.setValue(folly::makeUnexpected(ShardError::CRC_MISMATCH)); }
         return;
     }
 
-    auto shard = deserialize_shard(shard_msg);
-    switch (shard.info.state) {
-    case ShardInfo::State::OPEN: {
-        std::scoped_lock lock_guard(_flying_shard_lock);
-        auto iter = _flying_shards.find(lsn);
-        if (iter == _flying_shards.end()) {
-            LOGW("can not find flying shards on lsn {}", lsn);
-            if (ctx) { ctx->promise_.setValue(folly::makeUnexpected(ShardError::UNKNOWN)); }
-            return;
-        }
-        shard.chunk_id = iter->second.chunk_id;
-        _flying_shards.erase(iter);
-        // serialize the finalized shard msg;
-        shard_msg = serialize_shard(shard);
-        // persist the serialize result to homestore MetaBlkService;
-        homestore::hs()->meta_service().add_sub_sb(HSHomeObject::s_shard_info_sub_type,
-                                                   r_cast< const uint8_t* >(shard_msg.c_str()), shard_msg.size(),
-                                                   shard.metablk_cookie);
-        // update in-memory shard map;
-        do_commit_new_shard(shard);
+    if (crc32_ieee(init_crc32, value.bytes, value.size) != header.payload_crc) {
+        // header & value is inconsistent;
+        LOGW("replication message header is inconsistent with value, lsn:{}", lsn);
+        if (ctx) { ctx->promise_.setValue(folly::makeUnexpected(ShardError::CRC_MISMATCH)); }
+        return;
+    }
+
```

Comment:
if the header or value corrupted, the log can not be applied anymore, should we just crash ?

Reply:
actually I trend to use replace_pg() method to use a new pg to replace the existing corrupted pg in such case(shard replacement will be better if it is supported)